Add cloud code resolver example

185driver opened the pull request:
This PR is a response to a suggestion made in the [Parse Community Forum](https://community.parseplatform.org/t/adding-a-custom-graphql-mutation-to-parse-server/820/8?u=185driver) to add a custom Cloud Code resolver example to Parse's GraphQL docs.

I think the example fits the need for a real-life scenario showing the usefulness of custom resolvers and includes examples of both decoding and encoding a `Relay Global Object Identification`. 

The included example does **not** include a section on how to set up custom schema for use with a Cloud Code resolver on a Parse Server, as that is covered in the [README.md](https://github.com/parse-community/parse-server#customizing-your-graphql-schema) file. However, my personal view, having encountered difficulty in finding the setup instructions after much searching of the docs earlier, is that custom schema setup instructions for Cloud Code resolver use should be added to the docs as well. This would make it easier for newer users to discover and use this excellent feature of Parse Server.

Additionally, this PR fixes a minor punctuation error in the definition of a `Relay Global Object Identification`.

185driver committed:
Add cloud code resolver example

Moumouls commented:
I'll take a look tomorrow !

davimacedo reviewed (changes requested):
Thanks for adding this. It will help a lot. I just believe we should include this under the customization section.

185driver commented:
Sounds good. Shall I place it on the second level then (as a sibling with `Configuration`)?  Should I do it now or would you like me to wait for additional feedback to be given?

davimacedo commented:
Yes. I believe it makes sense. You can wait for @Moumouls feedback.

Moumouls commented on _includes/graphql/objects.md:
```diff
@@ -238,3 +238,95 @@ mutation aNestedMutation {
+  const cartItemId = Buffer.from(concatObjectId).toString("base64");
+
+  // Convert to a JSON object to handle adding the base64 id property.
+  const cartItemJson = savedCartItem.toJSON();
+  Object.assign(cartItemJson, { id: cartItemId });
+  return cartItemJson;
```
What you think about ES6 shortcuts
```suggestion
// Convert to a JSON object to handle adding the base64 id property.
  return {...savedCartItem.toJSON(), id: cartItemId }
```

185driver replied:
Very nice. Thank you.

Moumouls commented on _includes/graphql/objects.md:
````diff
@@ -238,3 +238,95 @@ mutation aNestedMutation {
+
+The ability to branch your resolver logic in this way enables you to replicate functionality found in Parse's auto-generated `createCartItem` and `updateCartItem` mutations and combine those behaviors into a single custom resolver.
+
+```sh
+# schema.graphql
+extend type Mutation {
+  addToCart(id: ID!): CartItem! @resolve
+}
+```
+
````
We need to show how to pass the schema into the ParseGraphQL Server, here developers can think that a simple `schema.graphql` at the root of the folder can do the job

185driver replied:
Agreed. I could use the example found in [this PR](https://github.com/parse-community/parse-server/pull/5821#issue-298733711). It's the code I used in my own project.

I'm hesitant to include the custom schema setup code in this section, as it seems to only be a related topic, and not quite this same one. Would it be okay for me to include the custom schema setup in the `Getting Started` section instead? If so, it could be inserted in one of a couple places:
1. In the `Using Express.Js` sub-section (mixed in with the other ParseServer setup code), or
2. Given its own sub-section named something like, `Adding Custom Schema` (I think I would prefer this method)

Moumouls commented on _includes/graphql/objects.md:
````diff
@@ -238,3 +238,95 @@ mutation aNestedMutation {
+}
+```
+
+**Note:** The `id` is a [Relay Global Object Identification](https://facebook.github.io/relay/graphql/objectidentification.htm); it's **not** a Parse `objectId`. Most of the time the `Relay Node Id` is a `Base64` of the `ParseClass` and the `objectId`. Cloud code does not recognize this type of `id`, so it must be converted to a Parse `objectId` for use in Cloud Code function queries.
+
````
To avoid some code like this
```
  const decoded = Buffer.from(id, "base64").toString();
  const itemObjectId = decoded.split(":")[1];
```

We need to inform developers that they can use `graphql-relay` lib.

May be add some reference about the graphql lib `graphql-relay` and the `const { type, id } = fromGlobalId(globalId);` method

Link: https://github.com/graphql/graphql-relay-js#object-identification
Lib: https://www.npmjs.com/package/graphql-relay

185driver replied:
Should I use my `Buffer` based code in the example, but add a text explanation and reference to the graphql lib as an optional solution, or should I dump the `Buffer` based code, use the graphql lib for the code example, and then reference the lib in the text?

With that lib available (thanks for mentioning it!) I personally would use it in my own projects, but there's something to be said for using code examples that are a bit more vanilla JS in the documentation too. Your preference?

Moumouls reviewed (commented):
Example how to pass schema (or ref to another part of the docs)
Little fixes

Moumouls commented:
@185driver  also may be move your code to `customisation.md` section :)

185driver committed:
Move cloud resolver code to customisation section

185driver committed:
Add custom schema setup guide

185driver committed:
Modify code example to use graphql-relay instead of Buffer

185driver commented:
I thought it might be easier to make the changes suggested by @Moumouls and submit them here for review. Hopefully things are moving in the right direction. Further suggestions are welcome.

Moumouls commented on _includes/graphql/customisation.md:
````diff
@@ -282,8 +282,106 @@ You can optionally override the default generated mutation names with aliases:
+
+The ability to branch your resolver logic in this way enables you to replicate functionality found in Parse's auto-generated `createCartItem` and `updateCartItem` mutations and combine those behaviors into a single custom resolver.
+
+```sh
+# schema.graphql
+extend type Mutation {
+  addToCart(id: ID!): CartItem! @resolve
+}
+```
+
````
Add a note/link to the getting started "Adding Custom Schema" section for better developer experience

185driver replied:
Will do, thanks for the reminder.

Moumouls reviewed (commented):
Seems really good, example with fromGlobalId is better than the base64 custom code.
Juste add some link about the schema.graphql and i think we are good to go !

185driver committed:
Improve custom schema and cloud code resolver docs

185driver commented:
I've added the link back to Getting Started and to the GraphQL Relay lib as suggested. 

After adding the custom schema section, the cloud code resolver section felt incomplete without a balancing query section, so I added that. Suggestions to change or remove it are welcome.

My feeling right now is that the two sections feel pretty complete and seem to fit together well. I believe the docs now provide someone like myself with enough official information to venture into Parse Server custom schema with a bit of confidence.

davimacedo reviewed (approved):
It looks pretty good to me. @Moumouls any additional change that you think we need here?

Moumouls commented:
Everything looks good to me :)

davimacedo commented:
Good job guys!
